sktime/decomposition/vampnet.py

```diff
@@ -0,0 +1,662 @@
+    c00_sqrt_inv = sym_inverse(c00, ret_sqrt=True, epsilon=epsilon, mode=mode)
+    ctt_sqrt_inv = sym_inverse(ctt, ret_sqrt=True, epsilon=epsilon, mode=mode)
+    return torch.chain_matmul(c00_sqrt_inv, c0t, ctt_sqrt_inv).t()
+
+
+def covariances(x: torch.Tensor, y: torch.Tensor, remove_mean: bool = True):
+    """Computes instantaneous and time-lagged covariances matrices.
+
+    Parameters
+    ----------
+    x : (T, n) torch.Tensor
+        Instantaneous data.
+    y : (T, n) torch.Tensor
+        Time-lagged data.
+    remove_mean: bool, default=True
+        Whether to remove the mean of x and y.
+
+    Returns
+    -------
+    cov_00 : (n, n) torch.Tensor
+        Auto-covariance matrix of x.
+    cov_0t : (n, n) torch.Tensor
+        Cross-covariance matrix of x and y.
+    cov_tt : (n, n) torch.Tensor
+        Auto-covariance matrix of y.
+
+    See Also
+    --------
+    sktime.covariance.Covariance : Estimator yielding these kind of covariance matrices based on raw data.
+    """
+
+    assert x.shape == y.shape, "x and y must be of same shape"
+    batch_size = x.shape[0]
+
+    if remove_mean:
+        x = x - x.mean(dim=0, keepdim=True)
+        y = y - y.mean(dim=0, keepdim=True)
+
+    # Calculate the cross-covariance
+    y_t = y.transpose(0, 1)
+    x_t = x.transpose(0, 1)
+    cov_01 = 1 / (batch_size - 1) * torch.matmul(x_t, y)
+    # Calculate the auto-correlations
+    cov_00 = 1 / (batch_size - 1) * torch.matmul(x_t, x)
+    cov_11 = 1 / (batch_size - 1) * torch.matmul(y_t, y)
+
+    return cov_00, cov_01, cov_11
+
+
+valid_score_methods = ('VAMP1', 'VAMP2', 'VAMPE')
+
+
+def score(data: torch.Tensor, data_lagged: torch.Tensor, method='VAMP2', epsilon: float = 1e-6, mode='trunc'):
+    r"""Computes the VAMP score based on data and corresponding time-shifted data.
+
+    Parameters
+    ----------
+    data : torch.Tensor
+        (N, d)-dimensional torch tensor
+    data_lagged : torch.Tensor
+        (N, k)-dimensional torch tensor
+    method : str, default='VAMP2'
+        The scoring method. See :meth:`score <sktime.decomposition.CovarianceKoopmanModel.score>` for details.
+    epsilon : float, default=1e-6
+        Cutoff parameter for small eigenvalues, alternatively regularization parameter.
+    mode : str, default='trunc'
+        Regularization mode for Hermetian inverse. See :meth:`sym_inverse`.
+
+    Returns
+    -------
+    score : torch.Tensor
+        The score.
+    """
+    if method not in valid_score_methods:
+        raise ValueError(f"Invalid method '{method}', supported are {valid_score_methods}")
+    assert data.shape == data_lagged.shape
+
+    if method == 'VAMP1':
+        koopman = koopman_matrix(data, data_lagged, epsilon=epsilon, mode=mode)
+        vamp_score = torch.norm(koopman, p='nuc')
+    elif method == 'VAMP2':
+        koopman = koopman_matrix(data, data_lagged, epsilon=epsilon, mode=mode)
+        vamp_score = torch.pow(torch.norm(koopman, p='fro'), 2)
```

Comment:
not that i know of anyone who uses vamp-n for n > 2, but it seems like this could be more general

Reply:
yes agreed, might add this later

Reply:
def no need to add now. to me it seems like the ideal would be to give an argument that's a number r or maybe `None` returns vamp E or something

Reply:
yes I agree and I thought about this too. Stuck to vamp 1, 2 and E for now because that is what was implemented in pyemma